### PR TITLE
openssh: add config option to disable passwords

### PR DESCRIPTION
**Pull request by pprindeville**
Maintainer: @tripolar 
Compile tested: x86_64, generic, OpenWrt HEAD (bcd17ce)
Run tested: same

Copied `.ipk` over to testbed.  Installed package and restarted server.

Inspected contents of `/etc/ssh/sshd_config`.  Disconnected and reconnected via ssh using public keys.

Description:

### CODE REVIEW

**Comment by tripolar** [comment]
you can just overwrite the config file. masking specific config options in the make file makes no sense to me

**Comment by pprindeville** [comment]
> you can just overwrite the config file. masking specific config options in the make file makes no sense to me

Not while you're building an image, you can't.

We added the recent limitation that `files/*` and the `base-files` package can overwrite a file provided by another package.

So there's no way to build a package or an image including a package (`openssh-server` in this case, obviously) which is inherently more secure.

Why do we need `CONFIG_BUSYBOX_CONFIG_FEATURE_SUID`?  Users could just get into the box and do `chmod u+s /bin/busybox`, right?

Why does `Package/lldpd/install` edit so many lines in `/etc/config/lldpd`?  Those could all be hand-edited once the user installs the file...

Also, we're doing conditionally what `Package/openssh-server-pam/install` does unconditionally.   Why not have users hand-edit the config file in that case?

The bottom line is (a) having the makefile customize a config file is nothing new (or even rare), and (b) proper security options should be made as "easy to get right" as humanly possible.

It's a lot easier for a user to figure out in `make menuconfig` "do I want to allow passwords for sshd or don't I" than it is for him to figure out, "what are the proper steps I need to go through to secure this service adequately" and have him read and understand the `sshd_config(5)` man page, etc.

When it comes to making a secure network device, we should make it as simple/stupid as possible.

This is a security issue: I propose that security issues be decided by a quorum among the core developers.

**Comment by tripolar** [comment]
based on your logic one has to mark dropbear too because it also allows to login with a password. so your argument makes no sense in my opinion. also disabling password logins by default is also unacceptable. i don't see a point in adding this option. can you point me to the patch that implements "We added the recent limitation that files/* and the base-files package can't overwrite a file provided by another package."

**Comment by pprindeville** [comment]
> based on your logic one has to mark dropbear too because it also allows to login with a password. 

And indeed, I was going to do that as a separate PR if this one was accepted.  I didn't want to spend time on **two** PR's only to see them both be shot down.

> also disabling password logins by default is also unacceptable.

A lot of distros do this, but it's trivial to change the default to `y`.  So I've done that.

> can you point me to the patch that implements "We added the recent limitation that files/* and the base-files package can't overwrite a file provided by another package."

I thought that files in `files/` couldn't overwrite files from a package or at least not from `base-files` but maybe I'm mistaken and confusing this with b044bd5.

**Comment by pprindeville** [comment]
> based on your logic one has to mark dropbear too because it also allows to login with a password.

Actually, that's already done:

```
config dropbear
	option PasswordAuth 'on'
	option RootPasswordAuth 'on'
	option Port         '22'
#	option BannerFile   '/etc/banner'
```

See lines 2 and 3.

**Comment by tripolar** [comment]
you can overwrite whatever you want with the "files/" folder
i'm sorry but i see this as half finished as it still doesn't create a working openssh when selecting the option, the user has to add files to have to possibility to login. if you want to create images like this, just use the overlay folder as this is not a standard config. porting all openssh options to uci would be a solution but i think that gets really complex but for shure possible

**Comment by bobafetthotmail** [comment]
I like the idea of standardizing this, but tripolar has a valid point, as-is this change just cripples the package and does not make its configuration easier.

If you add an option to disable the password on ssh you MUST add the logic to include the accepted keys in the firmware on build time too, and block compilation if no keys are found (so you won't flash a firmware with inaccessible ssh). This would be the foolproof way.

I don't know how this can be handled, maybe another (mandatory if the "nopassword" is selected) config option that takes a path where the keys are stored in the host system and copies them over in a /files directory?

**Comment by pprindeville** [comment]
Overlays don't work any more.  See https://github.com/lede-project/source/commit/b044bd5.

**Comment by pprindeville** [comment]
Sorry, I'm not sure how it's crippled.

I have a `files/root/.ssh/authorized_keys` file in my cloning, and I enable this option... and I get a more secure build.

Nothing is crippled.

**Comment by pprindeville** [comment]
> I don't know how this can be handled, maybe another (mandatory if the "nopassword" is selected) config option that takes a path where the keys are stored in the host system and copies them over in a /files directory?

I could work on this, but don't want to invest the time only to have it be nixed after the fact.  Is there consensus on doing it this way?

**Comment by bobafetthotmail** [comment]
> Sorry, I'm not sure how it's crippled.

There is no check that you actually have that file. If you don't have it and enable this option it would just make a crippled package.

As I said, make a check for that, or a way to specify the path to a custom file to use, and have it throw errors and fail compile if no such file is found.

**Comment by tripolar** [comment]
this seems to be code that overlays makefiles. and it happens at

https://github.com/lede-project/source/commit/b044bd5 - "define Build/IncludeOverlay"

Build time. Please read the code and don't point to commits that include buzzwords.

One also has to support this "extensions" that create no benefit compared to the "files/".
uci support would be the only clean solution

> Sorry, I'm not sure how it's crippled.
> 
> I have a files/root/.ssh/authorized_keys file in my cloning, and I enable this option... and I get a more secure build.
> 
> Nothing is crippled.

as i said just overwrite sshd_config with a file in files/etc/ssh/sshd_config

**Comment by bobafetthotmail** [comment]
> I could work on this, but don't want to invest the time only to have it be nixed after the fact. Is there consensus on doing it this way?

This is @tripolar 's call. I think that if you have checks in place to make sure that you can't accidentally compile a firmware without the whitelist of accepted keys (aka a broken ssh) it should be a good addition.

**Comment by tripolar** [comment]
i would prefer not supporting such an ugly hack (as there is already another ugly overlay hack in place ;) ). think about implementing **all** options as uci options. this would move this whole "intelligence" out of the makefile into the default image and every user would benefit from it

**Comment by pprindeville** [comment]
> Build time. Please read the code and don't point to commits that include buzzwords.

I've read the code.  When you wrote:

> if you want to create images like this, just use the overlay folder as this is not a standard config.

I thought you were talking about using `overlay/` at build-time.

**Comment by pprindeville** [comment]
> think about implementing all options as uci options.

How about starting out with a core set of the most useful ones, which you can call out if you want, and we can add others as need arises?  There are a lot of packages with many, many options (like ISC-DHCP) where we only support a small fraction of the total options and other things we either default or explicitly set to "sane" defaults.

**Comment by pprindeville** [comment]
We could, for instance, start with parity with `dropbear`'s options.

**Comment by pprindeville** [comment]
> As I said, make a check for that, or a way to specify the path to a custom file to use, and have it throw errors and fail compile if no such file is found.

@bobafetthotmail I don't know that there is a truly foolproof solution, BTW.

I could have the file `files/etc/config/dropbear`:

```
config dropbear
	option PasswordAuth 'off'
```

and not have any `files/root/.ssh/sshd_config` file, and still make an image for an unusable box.

**Comment by bobafetthotmail** [comment]
> You can create a broken files/ directory just as easily as you can create a broken .config

No, you can implement some form of sanity checks in makefiles reading the config, so it's harder to make a broken package that way.

Which is why I prefer it as a config option instead of just asking the builder to dump a whole config file and whitelisted keys in /files and integrating that with no check.

**Comment by pprindeville** [comment]
> No, you can implement some form of sanity checks in makefiles reading the config, so it's harder to make a broken package that way.

Packages are unaware of anything in `files/` by design.

**Comment by tripolar** [comment]
> How about starting out with a core set of the most useful ones, which you can call out if you want, and we can add others as need arises? There are a lot of packages with many, many options (like ISC-DHCP) where we only support a small fraction of the total options and other things we either default or explicitly set to "sane" defaults.

i think everything should be implemented and this was the main reason to not do it. lack of time and lack of time to check at every update which new options were added/removed.

i would prefer to stay with the old current version

**Comment by bobafetthotmail** [comment]
> Packages are unaware of anything in files/ by design. 

Which is why I said the makefile should handle this thing with a config option, without asking the user to place files in /files.

That way is unsafe by design, so there is nothing you can do about that.

**Comment by poranje** [comment]
> Op 29 jan. 2018, om 02:22 heeft tripolar <notifications@github.com> het volgende geschreven:
> 
> How about starting out with a core set of the most useful ones, which you can call out if you want, and we can add others as need arises? There are a lot of packages with many, many options (like ISC-DHCP) where we only support a small fraction of the total options and other things we either default or explicitly set to "sane" defaults.
> 
> i think everything should be implemented and this was the main reason to not do it. lack of time and lack of time to check at every update which new options were added/removed.
When "everything should be implemented" a lack of time is understandable, but why would that be ?
Think ? That's not a very convincing reason. Please elaborate a on the reason pro and against implementing a useful subset op the options in UCI.

When the problem is that the sshd_config could be overwritten by the init script whereby possibly destroying manually changed settings, then a setup such as in samba with an editable template might offer a solution.

**Comment by pprindeville** [comment]
See PR #6311.
